### PR TITLE
feat[notask]: add bci-whispercpp package for BCI neural signal transcription

### PR DESCRIPTION
**Pull request by sharmaraju352**
## Summary

- Adds `packages/bci-whispercpp` — a native C++ addon that transcribes brain-computer interface (BCI) neural signals into text using whisper.cpp
- Native GGML inference matches the Python BrainWhisperer reference output on 3/5 test samples (remaining 2 differ due to f16 quantization noise in the decoder)
- Mirrors the JS API surface of `@qvac/transcription-whispercpp` but replaces audio input with 512-channel neural signal input

## Architecture

```
neural signal (512ch) → NeuralProcessor (smooth, day projection)
  → whisper_set_mel (mel-major layout)
  → patched whisper.cpp encoder (windowed attention layers 0–3)
  → standard whisper decoder (beam search, LoRA-merged weights)
  → text
```

## Key components

| Component | Description |
|-----------|-------------|
| C++ addon | `BCIModel` + `NeuralProcessor` — Gaussian smoothing, day projection (from `bci-embedder.bin`), mel injection via `whisper_set_mel_with_state` |
| JS bindings | `BCIInterface` (batch + streaming), mirrors `AddonInterface` from transcription package |
| Model conversion | `scripts/convert-model.py` — PyTorch checkpoint → GGML model + embedder weights (LoRA merge, 6-layer encoder, conv1 k=7, windowed attention params, day projections) |
| vcpkg overlay | 4 patches to whisper.cpp: vcpkg build fix, Apple Silicon cross-compile, variable conv1 kernel, **windowed attention + SOS tokens** |
| Tests | Integration tests with WER measurement on 5 real neural signal samples |

## Model files

The conversion script produces **two files**, both required for inference:

| File | Size | Description |
|------|------|-------------|
| `ggml-bci-windowed.bin` | ~84 MB | GGML model: whisper encoder/decoder (LoRA-merged), tokenizer, positional embedding, windowed attention header |
| `bci-embedder.bin` | ~24 MB | Day projection weights: low-rank A·B matrices per recording day, month projections, session-to-day mapping |

Both files must be in the same directory at runtime. The addon fails with a clear error if the embedder is missing.

## Fixes applied for Python-matching output

| Fix | Description |
|-----|-------------|
| Mel layout transpose | Changed from frame-major to mel-major (`data[mel_bin * n_len + frame]`) to match whisper.cpp convention |
| Windowed attention | Patched whisper.cpp encoder to apply attention mask (window=57) on layers 0–3, matching Python's `build_window_mask` |
| Decoder SOS tokens | Force `[SOT, en, transcribe, notimestamps]` for BCI models even on English-only base (whisper-tiny.en skips `en`/`transcribe` by default) |
| Decoder params | `no_timestamps=true`, `single_segment=true`, `no_context=true` to match Python's generation config |
| Model header | Added `n_audio_window_size` and `n_audio_last_window_layer` fields after `n_audio_conv1_kernel` |

## Results (3/5 exact word match with Python reference)

| Sample | GGML native output | Python reference | Match |
|--------|--------------------|------------------|-------|
| 0 | "You can see the good at this point as well." | "you can see the good at this point as well" | ✅ |
| 1 | "How does it keep the cost down?" | "how does it keep the cost said" | ❌ |
| 2 | "Not too controversial." | "not too controversial" | ✅ |
| 3 | "The jury in a just work together on it." | "the jury and a judge work together on it" | ❌ |
| 4 | "We're quite vocal about it." | "we're quite vocal about it" | ✅ |

Average WER: 10.4% across all 5 samples. Samples 1 and 3 differ due to f16 quantization noise cascading through the transformer — the encoder outputs are close but small numerical differences change the decoder's beam search ranking.

## How to test

```bash
cd packages/bci-whispercpp
npm install
VCPKG_ROOT=/path/to/vcpkg npm run build

# Convert trained model (produces both ggml-bci-windowed.bin and bci-embedder.bin)
python3 scripts/convert-model.py \
  --checkpoint /path/to/epoch=93-val_wer=0.0910.ckpt

# Run integration tests
WHISPER_MODEL_PATH=./models/ggml-bci-windowed.bin npm run test:integration

# Run batch example
WHISPER_MODEL_PATH=./models/ggml-bci-windowed.bin npx bare examples/transcribe-neural.js --batch
```

## Test plan

- [x] Build succeeds on macOS arm64
- [x] Native GGML output matches Python reference (3/5 samples exact, 10.4% avg WER)
- [x] Integration tests pass (4/4 tests, 9/9 assertions)
- [x] Model conversion pipeline works end-to-end (produces both model files)
- [x] Addon fails with clear error when `bci-embedder.bin` is missing
- [ ] Verify on Linux x64
- [ ] Verify on Windows

## CI Workflows

- [Prebuild & Integration Tests](https://github.com/tetherto/qvac/actions/runs/24354984444)
- [Sanity Checks & Platform Prebuilds](https://github.com/tetherto/qvac/actions/runs/24354986715)

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: packages/bci-whispercpp/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "bci-whispercpp",
+  "version-string": "0.1.0",
+  "dependencies": [
+    {
+      "name": "qvac-lib-inference-addon-cpp",
+      "version>=": "1.1.2"
```
Comment (GustavoA1604) [inline packages/bci-whispercpp/vcpkg.json]: Use 1.1.5

File: packages/bci-whispercpp/vcpkg-overlays/whisper-cpp/0004-bci-windowed-attention.patch
Comment (GustavoA1604) [inline packages/bci-whispercpp/vcpkg-overlays/whisper-cpp/0004-bci-windowed-attention.patch:1]: This does not look like a patch file and is also not used by portfile.cmake. Need to check

File: packages/bci-whispercpp/index.js [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    if (opts.mode === 'onnx' && this._onnxConfig) {
+      return this._transcribeOnnx(filePath, opts)
+    }
+    const data = fs.readFileSync(filePath)
+    return this.transcribe(new Uint8Array(data))
+  }
+
+  /**
+   * Configure ONNX inference mode for Python-matching output.
+   * @param {Object} onnxConfig
```
Comment (GustavoA1604) [inline packages/bci-whispercpp/index.js]: Could you elucidate how is it using onnx if we are using whisper-cpp backend?

File: packages/bci-whispercpp/addon/src/addon/AddonJs.hpp
Comment (GustavoA1604) [inline packages/bci-whispercpp/addon/src/addon/AddonJs.hpp:1]: Missing AddonCpp.hpp?

File: packages/bci-whispercpp/scripts/infer.py
Comment (GustavoA1604) [inline packages/bci-whispercpp/scripts/infer.py:1]: Probably this file is not meant to be added?

File: packages/bci-whispercpp/scripts/patch-ggml-model.py
Comment (GustavoA1604) [inline packages/bci-whispercpp/scripts/patch-ggml-model.py:1]: Why is it not part of convert-model.py?

File: packages/bci-whispercpp/test/fixtures/python_predictions.json
Comment (GustavoA1604) [inline packages/bci-whispercpp/test/fixtures/python_predictions.json:1]: Probably adding some files that are not meant to be added? python_predictions, onnx-compare.js, etc.

File: packages/bci-whispercpp/test/integration/bci-addon.test.js [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+  const samplePath = getSamplePath(sample.file)
+  if (!fs.existsSync(samplePath)) {
+    t.skip(`Sample file missing: ${samplePath}`)
+    return
+  }
+
+  const segments = []
+  let stats = null
+
+  const onOutput = (addon, event, jobId, data, error) => {
```
Comment (GustavoA1604) [inline packages/bci-whispercpp/test/integration/bci-addon.test.js]: Shouldnt this be part of addon callback in index.js?
Integration tests can rely on package-level interface instead of binding-level interface

File: packages/bci-whispercpp/bci.js [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+          this._setState(state.LISTENING)
+          throw err
+        }
+        if (!accepted) {
+          this._setState(state.LISTENING)
+          throw new Error('Cannot set new job: a job is already set or being processed')
+        }
+
+        this._activeJobId = currentJobId
+        this._nextJobId += 1
```
Comment (GustavoA1604) [inline packages/bci-whispercpp/bci.js:200]: Right now whisper, parakeet and TTS are handling jobID in JS level, but might be better and less error-prone to do like llamacpp-llm and simplify this logic, since BCI would mostly be used with streaming capabilities anyway

File: packages/bci-whispercpp/README.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
+# @qvac/bci-whispercpp
+
+Brain-Computer Interface (BCI) neural signal transcription addon for qvac, powered by [whisper.cpp](https://github.com/ggerganov/whisper.cpp).
```
Comment (GustavoA1604) [inline packages/bci-whispercpp/README.md]: Should point to our fork

**Review by GustavoA1604 — CHANGES_REQUESTED** [review]

**Comment by ogad-tether** [comment]
Prebuilds and mobile e2e failing

**Comment by sharmaraju352** [comment]
## This PR has been split into 6 stacked PRs for easier review

The original 6,309-line PR has been broken up by concern:

| # | Ticket | PR | Lines | Focus |
|---|--------|-----|------:|-------|
| 1 | QVAC-17057 | #1583 | ~4,200 | POC: C++ addon + build + JS batch API + tests |
| 2 | QVAC-17071 | #1584 | ~130 | Switch whisper-cpp overlay to tetherto fork |
| 3 | QVAC-17072 | #1585 | ~370 | Consume whisper-cpp from registry, remove overlays |
| 4 | QVAC-17062 | #1586 | ~130 | Add streaming transcription + full integration tests |
| 5 | QVAC-17058 | #1587 | ~50 | Empty BCI workflows (enable branch triggering) |
| 6 | QVAC-17063 | #1588 | ~2,000 | Full CI workflows + mobile test code |

Each PR is stacked on the previous one. Please review starting from #1583.

Closing this PR in favor of the stack.
